messaging/broker.go

```diff
@@ -144,943 +222,944 @@ func (b *Broker) closeTopics() {
+	if len(segments) == 0 {
+		return 0, nil
 	}
 
-	// Stream each topic sequentially.
-	for _, t := range hdr.Topics {
-		if _, err := copyFileN(w, t.path, t.Size); err != nil {
-			return 0, err
-		}
+	// Read highest index on the last segment.
+	index, err := ReadSegmentMaxIndex(segments.Last().Path)
+	if err != nil {
+		return 0, fmt.Errorf("read segment max index: %s", err)
 	}
 
-	// Return the snapshot and its last applied index.
-	return hdr.maxIndex(), nil
+	return index, nil
 }
 
-// Restore reads the broker state.
-func (fsm *brokerFSM) Restore(r io.Reader) error {
-	b := (*Broker)(fsm)
-
-	b.mu.Lock()
-	defer b.mu.Unlock()
-
-	// Read header frame.
-	var sz uint32
-	if err := binary.Read(r, binary.BigEndian, &sz); err != nil {
-		return fmt.Errorf("read header size: %s", err)
-	}
-	buf := make([]byte, sz)
-	if _, err := io.ReadFull(r, buf); err != nil {
-		return fmt.Errorf("read header: %s", err)
-	}
+// WriteMessage writes a message to the end of the topic.
+func (t *Topic) WriteMessage(m *Message) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
 
-	// Decode header.
-	s := &snapshotHeader{}
-	if err := json.Unmarshal(buf, &s); err != nil {
-		return fmt.Errorf("decode header: %s", err)
+	// Return error if message index is lower than the topic's highest index.
+	if m.Index <= t.index {
+		return ErrStaleWrite
 	}
 
-	// Close any topics and replicas which might be open and clear them out.
-	b.closeTopics()
-	b.closeReplicas()
-
-	// Copy topic files from snapshot to local disk.
-	for _, st := range s.Topics {
-		t := b.createTopic(st.ID)
-		t.index = st.Index
-
-		// Remove existing file if it exists.
-		if err := os.Remove(t.path); err != nil && !os.IsNotExist(err) {
-			return err
-		}
-
-		// Open new empty topic file.
-		if err := t.open(); err != nil {
-			return fmt.Errorf("open topic: %s", err)
+	// Close the current file handle if it's too large.
+	if t.file != nil {
+		if fi, err := t.file.Stat(); err != nil {
+			return fmt.Errorf("stat: %s", err)
+		} else if fi.Size() > t.MaxSegmentSize {
+			_ = t.file.Close()
+			t.file = nil
 		}
+	}
 
-		// Copy data from snapshot into file.
-		if _, err := io.CopyN(t.file, r, st.Size); err != nil {
-			return fmt.Errorf("copy topic: %s", err)
+	// Create a new segment if we have no handle.
+	if t.file == nil {
+		f, err := os.OpenFile(t.segmentPath(m.Index), os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0600)
+		if err != nil {
+			return fmt.Errorf("create segment file: %s", err)
 		}
+		t.file = f
 	}
 
-	// Update the replicas.
-	for _, sr := range s.Replicas {
-		// Create replica.
-		r := newReplica(b, sr.ID, sr.URL)
-		b.replicas[r.id] = r
+	// Encode message.
+	b := make([]byte, messageHeaderSize+len(m.Data))
+	copy(b, m.marshalHeader())
+	copy(b[messageHeaderSize:], m.Data)
 
-		// Append replica's topics.
-		for _, srt := range sr.Topics {
-			r.topics[srt.TopicID] = srt.Index
-		}
+	// Write to last segment.
+	if _, err := t.file.Write(b); err != nil {
```

Comment:
We don't have recovery support for partial writes in this PR. @pauldix has a list of issues to fix beyond this PR. That should be one of them IIRC.